### ISSUE TITLE
Neo-Elements disappear after saving

### ISSUE THREAD

**Issue by seichenberger** [spicywebau/craft-neo#120]
We have build a site for our customer largely based on your faboulus plugin. After the Neo Plugin update to version 1.5 we see strange things happening (a downgrade to 1.4.1 solved these problems):

- The lightswitch problem as mentionend in https://github.com/benjamminf/craft-neo/issues/118 occurs
- On a page with 3 Neo matrices: On one matrix, it is not possible to add an Element after loading the page. After saving the page, it works (similar to the lightswitch problem). After adding another element in another matrix on this page and saving the page, the previously added elements to the first matrix disappear.
- In another matrix, the addition of two elements worked well. After some further changes to other matrices the elements are shown doubled in the frontend, but only one of these elements is visible in the backend.
- The removal of a matrix in settings > section > entry types did not work properly: The elements were still visible in the frontend after saving the page without the corresponding matrix.

Probably the minimal requirements for reproducing the problem: Page with 2 Neo matrices. Adding elements to the first matrix. Save the page. Add another Element to the other matrix. Save. If the first element is not present anymore, the bug has been reproduced.

We tried to fix the issues by rebuilding the page: Removal of the Neo matrices and adding them again. This did not solve the problem.

We assume that there is a problem with updating the database. We are highly interested in a bugfix and offer the best we can for supporting you.

Addendum: Initially we stated that the problem occured after the Craft CMS update to version 2.6.2997. We found that it was the Neo Plugin update that caused the problem. We now use Neo 1.4.1 and Craft 2.6.2997 and the stated problems do not occure anymore.

**Comment by dennisfrank** [comment]
Unfortunately, I am also struggling with both issues. (#118 is the minor one.) Losing content when saving an entry could be catastrophic.

All Matrix-in-Neo content get erased when editing another Matrix or Neo field.

**Comment by benjamminf** [comment]
Hey guys, thanks for reporting. I'm well aware of the issue, which stems from a feature I added in 1.5.0 for copying content to other locales. I'm going to bring that version down in the meantime while I work on a patch for this.

Downgrading to 1.4.1 should get things back in order. There were no database changes between those versions so it's safe to do so.

**Comment by madebyfir3** [comment]
There are also issues with the new "copy & paste" changed from "duplicate" where there are two neo fields on a template, remove and unorder the other neo fields lets say in a sidebar, weird. Backing up to version with duplicate neo fields fixes the issue.

**Comment by knynkwl** [comment]
@benjamminf Looks like 1.5.0 is still up downloadable, were you going pause that version and revert to 1.4? Just worring about other people having this bug. I am having a lot of issues with 1.5

**Comment by knynkwl** [comment]
This is another that issue I was having with 1.5 https://github.com/benjamminf/craft-neo/issues/118

**Comment by terryupton** [comment]
Hey @benjamminf I am also experiencing this issue and #118 - would we best to downgrade? I can only see 1.5 available. Thanks.

**Comment by svenpeeters** [comment]
Any movement on this?

**Comment by seamusmh** [comment]
FYI, if you're having this issue, try increasing max_input_var in PHP. The default is 1000, and if you have a Neo setup with lots of layout options in every row, it's pretty easy to get up over 1000 on a long post. At least that was why my fields were getting truncated. No warning in Craft about it either.

**Comment by justenh** [comment]
We just reverted our neo plugin to 1.4. I'm not sure if it fixed the issue. We'll just have to wait and see.

Posting this here as a warning to anyone else who is encountering this issue as of today. I must have downloaded the plugin just before it was taken down.

**Comment by jerwarren** [comment]
I'm having this issue with the version currently available here on github (which, incidentally, claims to be version 1.5.0 in the plugin settings.)

**Comment by SuperlabTina** [comment]
Is there any update on this matter?

**Comment by ritagya-meharishi** [comment]
Thanks @seamusmh, do you know if I can make this change from Control Panel?

**Comment by seamusmh** [comment]
No. It needs to be done in the php.ini file.

**Comment by caferelgin** [comment]
Any news with this bug? Is there a workaround?

**Comment by rbaud** [comment]
**Warning: this issue, from well over a year ago, still seems to be occurring with the most-recent commit (11/29/18) to the craft-2 branch.**

In this version, and the now-pulled 1.5.0, if 2 Neo fields are present on an Entry and populated with any configuration of blocks, adding any block (by either manual addition or copy/paste) to the 1st field will utterly destroy the 2nd field.

@ttempleton, this is a critical issue that can ruin entire sites, and it's preventing us from upgrading from 1.4.1, which means we have to go without the much-needed copy/paste functionality. Is there a fix coming down the pike for this?

**Comment by jerwarren** [comment]
I struggled a bunch with this.  I've found that even with the fixed release (which can be downloaded from the releases page on github instead of pulling from git) it's possible for an entry to get into a broken state in what I've determined is a race condition due to multiple people saving the same entry at the same time or flipping between various revisions of a page. 

If you're seeing this on fresh new pages that haven't previously had neo blocks, then that's different, but I suspect that you, like me, didn't actually have the fixed release because you didn't get it from the releases page.  Once you DO have the fixed release, it's not going to magically fix the broken neo blocks you already have.  I had to go into the database and delete some "orphaned" blocks that were no longer associated with the entry in question.

Basically what happens is that the "missing" blocks are there in the database, but are not correctly associated with the neo block that's in the affected entry.  It seems to get confused until they're deleted.

In any case, I suggest grabbing the release from the releases page, make a new page with a new neo block and see if still behaves the way you describe it.  I suspect it won't, and then it's just a matter of cleaning up the affected neo blocks.

**Comment by rbaud** [comment]
@jerwarren thanks for the feedback!

I'm working on a local VM, so I'm sure the manifestation of this issue I mentioned [above](https://github.com/spicywebau/craft-neo/issues/120#issuecomment-451472558) isn't caused by a race condition with multiple editors/admins.

I've also confirmed that both the 1.5.0 release (which was pulled from the releases page in Nov 2017) and the craft-2 branch at the latest commit ([653b632](https://github.com/spicywebau/craft-neo/commit/653b632bc000faa911eb812e0545abc8bcad11eb), 11/28/18) are affected. Following your advice, I cleared the `craft_neoblocks` table, deleted my Neo fields, and created 2 fresh Neo fields for my test Entry. Unfortunately, the problem persists.

I did a bit of digging in the code, fiddling with this 2 Neo field setup, and I think I've narrowed down the issue to, like you mentioned, extra rows being added to `craft_neoblocks`. I'm reasonably certain that this an issue revolving around field ids and using them to determine if a element being saved is new or not.

Although I'm still testing, it seems that changing line 48 in the `init()` function of neo/fieldtypes/NeoFieldType.php to check for the existence of `elementId` instead of `id` seems to fix the problem. See the full `init()` function below and note the change to `$this->_isNewElement = ...`

```
        /**
	 * Initialisation of the field type instance.
	 * Used for binding events.
	 */
	public function init()
	{
		parent::init();

		// Watch to see if this request is saving an element
		// This is so that keywords can be safely offloaded to a task later on
		craft()->on('elements.onBeforeSaveElement', function($e)
		{
			$this->_savingElement = $e->params['element'];
			// $this->_isNewElement = !((bool) $this->_savingElement->id);
			$this->_isNewElement = isset($this->_savingElement->elementId);
		});
	}
```

With this change (to the current craft-2 branch), multiple Neo fields work properly again, there are no errant rows added to the database, and copy/paste (inside a Neo field and between 2 versions of the same field on different Entries) works too. More testing to do I'm sure, but so far so good.

**Comment by jerwarren** [comment]
Oh! That reminds me of something I forgot:  

There's a config for whether neo saves everything or tries to save only changed fields.  Setting that to always save everything is when I think everything started working OK for me (save for the race condition.)

I think your code change probably achieves something similar.

**Comment by rbaud** [comment]
@jerwarren thanks again for following up. Continued testing has proven that the change I posted [above](https://github.com/spicywebau/craft-neo/issues/120#issuecomment-451525969) does not fully work. Sometimes -- I can't quite pin it down -- the 2nd Neo field will become corrupted (due to multiple rows being created in `craft_neoblocks` for the same block).

However, your suggestion to toggle the config setting seems to be working! I'll continue to test that, and I'll post again if I run into issues. But for the time being, anyone who is seeing this issue should try toggling `saveModifiedBlocksOnly` in **craft/config/neo.php** to false.

Like this:
```
<?
  return [
    // If true, optimizes the saving of Neo fields by only saving existing blocks that have had their content modified
    'saveModifiedBlocksOnly' => false,
...
```

I think any potential performance hit is well worth sidestepping this critical bug.

**Comment by ttempleton** [comment]
@caferelgin, @rbaud, @jerwarren apologies for the delay with work on Neo for Craft 2 since late November.

As you've discussed, this does involve extra rows being added to the `neoblocks` table.  Specifically, when adding new blocks to existing fields, I notice that only the new block will be saved for the first field, but it appears all blocks are being saved as new blocks for the other fields -- while retaining the existing blocks in the `neoblocks` table.  The block structures are then regenerated -- as they always are when resaving a field -- but for the affected fields, that process seems to only take into account the 'new' blocks, ignoring the previously existing blocks.  This then causes most of those blocks not to appear, because the block queries being performed are finding the first of the original blocks which is no longer associated with any block structure.  Deleting the orphaned blocks, as per @jerwarren's suggestion, does return the situation to normal.

As @rbaud has said, this suggests an issue with checking whether or not a block is new on entries with more than one Neo field.  Unfortunately, though @jerwarren's suggestion of disabling the `saveModifiedBlocksOnly` setting would make sense, it doesn't appear to be working for me.  I'll continue to look into this; both of you are welcome to continue looking into this as well, if you'd like, the extra efforts to track down the exact cause would definitely be appreciated.

That all being said, the former Neo release 1.5.0 was unstable, due to bugs like this and #118 -- hence why it was pulled from the release feed, and therefore should not be used in production environments.  Anyone wanting to use the latest from the `craft-2` branch to help test any bugfixes for the 1.5.1 release in a development environment is welcome to do so, but for anyone else, until Neo 1.5.1 is released, the recommended course of action remains to use the official most recent stable release of the Neo 1.x series, which is 1.4.1.

**Comment by ttempleton** [comment]
The issue is that the field type `_savingElement` and `_isNewElement` properties are being set before saving *every* element -- things like Neo and Matrix blocks included -- when `_isNewElement` should only be set for the overall element being saved, e.g. an entry or category.  I've changed it to only set `_savingElement` and therefore `_isNewElement` for the first element, which will be the entry/category/etc.  So far, this change appears to be working well.  I won't close this issue yet, though, just in case, and will continue to test in different situations.

**Comment by rbaud** [comment]
@jerwarren, thanks for following up again!

And thanks, @ttempleton, for looking into this! Excited to here an officially-fixed release may be coming for this. The effort is very much appreciated! Thank you.

**Comment by theskyfloor** [comment]
I seem to have this issue but going back to 1.4.1 has't helped with my bug which seems to be related to the table maker plug-in? Any chance that could be confirmed? Have a client site where they are struck by this bug and the site editing has had to grind to a halt

**Comment by theskyfloor** [comment]
Actually it does it for me no matter what. Even on 1.4.1... any ideas?

**Comment by theskyfloor** [comment]
Actually still happening when I try 1.5.1 and no matter what fields I try : (

**Comment by ttempleton** [comment]
@theskyfloor I've just been testing with Table Maker and yet to run into any problems.  This issue definitely should not occur in Neo 1.4.1 or 1.5.1, though, so I think there's something else going on.  Just in case, can you please check your `neoblocks` database table after saving a field and confirm whether it does have the same situation as described in [this comment](https://github.com/spicywebau/craft-neo/issues/120#issuecomment-451651045)?

**Comment by theskyfloor** [comment]
Will check!

**Comment by ttempleton** [comment]
@theskyfloor, have you had a chance to check whether you had the extra rows being added to the `neoblocks` table with Neo 1.4.1 or 1.5.1?

**Comment by theskyfloor** [comment]
Unfortunately the client changed the login creds for the DB and they haven't been able to provide me with updated ones :(

**Comment by ttempleton** [comment]
No worries, hopefully it's not too long before they can provide you with those.  I suspect at the moment that you were just experiencing issue #128, since that issue can present in a similar way to this one (if the newly-added block does not have any allowed child block types, in my testing), but this issue could still be involved if the site was running Neo 1.5.0 at the time the problem appeared.

**Comment by ttempleton** [comment]
After further testing and improvements, we feel confident enough to release Neo 1.5.1 now with the fix for this issue.  Due to the move to the `craft-2` branch since 1.5.0, this release won't appear on a Craft installation's updates page, so it'll need to be downloaded from its [release page on GitHub](https://github.com/spicywebau/craft-neo/releases/tag/1.5.1).

@theskyfloor if you regain access to that client's site, try out the 1.5.1 release and let us know if the situation has been resolved.

**Comment by engram-design** [comment]
Not sure if I'm experiencing this same issue or not. Recently upgraded site from Craft 2 to Craft 3. Neo blocks were intact, but when I go to save, all blocks disappear. I can see the blocks exist in the db table, linked to the correct owner, but they don't appear.

If I switch back to an earlier revision, the blocks are there, but when trying to revert, the Neo field remains empty. Anyone with a similar issue for Craft 3?

**Comment by ttempleton** [comment]
@engram-design Which versions of Neo for Craft 2/3 were you using?  I don't think it's related to this issue, but it sounds like something might have gone wrong with the block structure somehow.  Could you please check the `neoblockstructures` table for the `structureId` for an affected field/owner/site, and then check the `structureelements` table for the elements with that `structureId` and see if their `elementId`s match the IDs of the Neo blocks?

**Comment by engram-design** [comment]
We were using 1.4.1 for C2, and 2.2.5 for C3. At first, I thought it might be because all the `craft_neoblocks` and `craft_neoblockstructures` rows had `NULL` as the `ownerSiteId`, but setting that to 1 didn't seem to help.

I've cross-referenced all those tables/values and everything seems to be in place. Going to clone this site locally and dive in...

**Comment by engram-design** [comment]
@ttempleton Yep, ignore this - this is an issue with https://github.com/aelvan/Preparse-Field-Craft, as disabling the plugin rectifies the situation. However, a handy thing to note in your support.

**Comment by ttempleton** [comment]
@engram-design Will do, thanks!

**Comment by evanwarner** [comment]
@ttempleton I don't think this is fixed, even in 1.5.2. Working on a site, I'm still loading up pages in Craft that suddenly have half their Neo blocks missing. The blocks are still in the database, and they seem to render on the front-end still. But if I want to edit the page, I need to recreate all the content because as soon as it's saved, the orphaned blocks are purged from the database.

**Comment by jerwarren** [comment]
I'm still seeing this happen on 1.5.2 with the latest craft version.

I seem to be losing fields randomly, and it SEEMS as if it's happening on fields that were previously removed from the page.

How can I best help troubleshoot?

**Comment by kblizeck** [comment]
Also experiencing issues with Neo 1.5.0 deleting fields on save. May be related to #128 - I tinkered locally with my database and deleted duplicate entries in `neoblockstructures` for the `ownerId` that was erroring on save, and the fields are appearing as expected. Definitely not sure what the repercussions are for deleting those records though...
Also, increased `max_input_vars=5000`

**Comment by twosixcode** [comment]
> @ttempleton Yep, ignore this - this is an issue with https://github.com/aelvan/Preparse-Field-Craft, as disabling the plugin rectifies the situation. However, a handy thing to note in your support.

In case it helps anyone else I'll just add here that we ran into this, too. Some neo blocks were being lost for us on a Craft 2 to 3 update. But taking Preparse out of the equation solved that issue for us.

**Comment by tekstrand** [comment]
Can confirm we're seeing in 1.5.2 as well, will update with a fix if I find anything.

**Comment by rhoffmann** [comment]
I have this problem appearing again after updating to the latest version of the neo plugin on a multi language site. (Craft 3.4.13 / Neo 2.7.5) When saving any entry (basically creating a new version) in a locale, the neo content on the other locales just disappears. Resetting it to an old version of the entry did not work.

I remember that I had this issue also a few months back, it cost me a lot of time to figure this out, and i think i somehow 'fixed' the issue by saving each page again as a new entry. I would loose the old versions, but somehow the neo entries reappeared. Is there any permanent solution to this issue, or is a workaround known? Currently it is quite impossible to do a system update without breaking the whole site, and not even knowing because the update process went 'ok'.

**Comment by pvldigital** [comment]
Hey @rhoffmann,

Can you update to the latest neo version? There has been some changes when saving for multisites and how the structure is created for it.

**Comment by rhoffmann** [comment]
Hello @pvldigital,
thanks for the hint. I updated to this version and it seems to have fixed it, i could revert the entry to a previous state.
